File: quartz/src/main/java/com/vitorpamplona/quartz/nip03Timestamp/ots/http/Response.java
package com.vitorpamplona.quartz.nip03Timestamp.ots.http;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.json.JsonMapper;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;

/**
 * Holds the response from an HTTP request.
 */
public class Response {
    private InputStream stream;
    private String fromUrl;
    private Integer status;

    public Response() {
    }

    public Response(InputStream stream) {
        this.stream = stream;
    }

    public void setStream(InputStream stream) {
        this.stream = stream;
    }

    public Integer getStatus() {
        return status;
    }

    public void setStatus(Integer status) {
        this.status = status;
    }

    public boolean isOk() {
        return getStatus() != null && 200 == getStatus();
    }

    public String getFromUrl() {
        return fromUrl;
    }

    public void setFromUrl(String fromUrl) {
        this.fromUrl = fromUrl;
    }

    public InputStream getStream() {
        return this.stream;
    }

    public String getString() throws IOException {
        return new String(getBytes(), StandardCharsets.UTF_8);
    }

    public byte[] getBytes() throws IOException {
        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        int nRead;
        byte[] data = new byte[16384];

        while ((nRead = this.stream.read(data, 0, data.length)) != -1) {
            buffer.write(data, 0, nRead);
        }

        buffer.flush();

        return buffer.toByteArray();
    }

    public JsonNode getJson() throws IOException {
        String jsonString = getString();
        JsonMapper builder = JsonMapper.builder().build();
        return builder.readTree(jsonString);
    }
}

File: quartz/src/androidMain/kotlin/com/vitorpamplona/quartz/nip03Timestamp/ots/http/Response.kt
/**
 * Copyright (c) 2025 Vitor Pamplona
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
 * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.vitorpamplona.quartz.nip03Timestamp.ots.http

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.json.JsonMapper
import java.io.ByteArrayOutputStream
import java.io.Closeable
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets

/**
 * Holds the response from an HTTP request.
 */
class Response : Closeable {
    private var stream: InputStream? = null

    var fromUrl: String? = null
    var status: Int? = null

    constructor()

    constructor(stream: InputStream) {
        this.stream = stream
    }

    fun setStream(stream: InputStream) {
        this.stream = stream
    }

    val isOk: Boolean
        get() = this.status != null && 200 == this.status

    fun getStream(): InputStream = this.stream!!

    @get:Throws(IOException::class)
    val string: String
        get() = kotlin.text.String(this.bytes, StandardCharsets.UTF_8)

    @get:Throws(IOException::class)
    val bytes: ByteArray
        get() {
            val buffer = ByteArrayOutputStream()
            var nRead: Int
            val data = ByteArray(16384)

            while ((this.stream!!.read(data, 0, data.size).also { nRead = it }) != -1) {
                buffer.write(data, 0, nRead)
            }

            buffer.flush()

            return buffer.toByteArray()
        }

    @get:Throws(IOException::class)
    val json: JsonNode?
        get() {
            val jsonString = this.string
            val builder =
                JsonMapper.builder().build()
            return builder.readTree(jsonString)
        }

    override fun close() {
        stream?.close()
        stream = null
    }
}
